cloud/terraform/machineactuator.go

```diff
@@ -351,10 +353,119 @@ func (tf *TerraformClient) PostDelete(cluster *clusterv1.Cluster, machines []*cl
+			}
+		} else {
+			glog.Info("UPDATE TYPE NOT SUPPORTED")
+			return nil
+		}
+	} else {
+		glog.Info("NODE UPDATES NOT SUPPORTED")
+		return nil
+	}
+
 	return nil
 }
 
+// Assumes that update is needed.
+// For now support only K8s control plane upgrades.
+func (tf *TerraformClient) updateMasterInPlace(goalMachine *clusterv1.Machine) error {
+	goalMachineControlPlaneVersion := goalMachine.Spec.Versions.ControlPlane
+	goalMachineKubeletVersion := goalMachine.Spec.Versions.Kubelet
+
+	currentControlPlaneVersion := goalMachine.Annotations[ControlPlaneVersionAnnotationKey]
+	currentKubeletVersion := goalMachine.Annotations[KubeletVersionAnnotationKey]
+
+	// Control plane upgrade
+	if goalMachineControlPlaneVersion != currentControlPlaneVersion {
+		// Pull the latest kudeadm
+		cmd := "export KUBEADM_VERSION=$(curl -sSL https://dl.k8s.io/release/stable.txt); " +
+			"curl -sSL https://dl.k8s.io/release/${KUBEADM_VERSION}/bin/linux/amd64/kubeadm | sudo tee /usr/bin/kubeadm > /dev/null; " +
+			"sudo chmod a+rx /usr/bin/kubeadm"
+		_, err := tf.remoteSshCommand(goalMachine, cmd, "~/.ssh/id_rsa", "ubuntu")
+		if err != nil {
+			glog.Infof("remoteSshCommand failed while downloading new kubeadm: %+v", err)
+			return err
+		}
+
+		// Next upgrade control plane
+		cmd = fmt.Sprintf("sudo kubeadm upgrade apply %s -y", "v"+goalMachine.Spec.Versions.ControlPlane)
+		_, err = tf.remoteSshCommand(goalMachine, cmd, "~/.ssh/id_rsa", "ubuntu")
+		if err != nil {
+			glog.Infof("remoteSshCommand failed while upgrading control plane: %+v", err)
+			return err
+		}
+	}
+
+	// Upgrade kubelet
+	if goalMachineKubeletVersion != currentKubeletVersion {
+		cmd := fmt.Sprintf("sudo kubectl drain %s --kubeconfig /etc/kubernetes/admin.conf --ignore-daemonsets", goalMachine.Name)
+		tf.remoteSshCommand(goalMachine, cmd, "~/.ssh/id_rsa", "ubuntu")
+
+		// Upgrade kubelet to desired version.
+		cmd = fmt.Sprintf("sudo apt-get install kubelet=%s", goalMachine.Spec.Versions.Kubelet+"-00")
+		_, err := tf.remoteSshCommand(goalMachine, cmd, "~/.ssh/id_rsa", "ubuntu")
+		if err != nil {
+			glog.Infof("remoteSshCommand while installing new kubelet version: %v", err)
+			return err
+		}
+
+		// Mark node as schedulable
+		cmd = fmt.Sprintf("sudo kubectl uncordon %s --kubeconfig /etc/kubernetes/admin.conf", goalMachine.Name)
+		_, err = tf.remoteSshCommand(goalMachine, cmd, "~/.ssh/id_rsa", "ubuntu")
+		if err != nil {
+			glog.Infof("remoteSshCommand while marking node as schedulable: %v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (tf *TerraformClient) remoteSshCommand(m *clusterv1.Machine, cmd, privateKeyPath, sshUser string) (string, error) {
+	glog.Infof("Remote SSH execution '%s' on %s", cmd, m.ObjectMeta.Name)
+
+	publicIP, err := tf.GetIP(m)
+	if err != nil {
+		return "", err
+	}
+
+	command := fmt.Sprintf("echo STARTFILE; %s", cmd)
+	c := exec.Command("ssh", "-i", privateKeyPath, sshUser+"@"+publicIP, command)
```

Comment:
`-o StrictHostKeyChecking no -o UserKnownHostsFile /dev/null`

Also adding `-q` and separating error and stdout should let you get rid of the `echo STARTFILE` hack.

Reply:
Done

Reply:
Did the `-q` and only reading stdout not work?